### ISSUE TITLE
conftest.py files in packages matching "test*" are being loaded, even if they're not in the `testpaths` list

### ISSUE THREAD

**Issue by mattalexx** [pytest-dev/pytest#9469]
- [ ] a detailed description of the bug or problem you are having
- [ ] output of `pip list` from the virtual environment you are using
- [ ] pytest and operating system versions
- [ ] minimal example if possible

I am defining `testpaths = tests` in `pytest.ini` and that's being picked up by pytest (see pytest output below), but then it's loading a `conftest.py` from another directory.

## Reproduction steps

The test files:

```shell
% tree
.
├── pytest.ini
├── tests
│   ├── __init__.py
│   └── test_.py
└── testsinvalid
    └── conftest.py
```

Their contents:

```shell
% find -type f | xargs -IF sh -c "echo; echo === F; cat F"

=== ./pytest.ini
[pytest]
testpaths = tests

=== ./testsinvalid/conftest.py
raise Exception('Loading %s!' % __file__)

=== ./tests/test_.py
print('Loading %s' % __file__)
def test_p():
    assert True

=== ./tests/__init__.py
print('Loading %s' % __file__)
```

The command to run:

```
PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 pytest --capture=no
```

## What I expect to see

```
% PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 pytest --capture=no
=================================================================== test session starts ====================================================================
platform linux -- Python 3.10.1, pytest-6.2.5, py-1.11.0, pluggy-0.13.1
rootdir: /[snip], configfile: pytest.ini, testpaths: tests
collecting ... Loading /[snip]/tests/__init__.py
Loading /[snip]/tests/test_.py
collected 1 item

tests/test_.py .

==================================================================== 1 passed in 0.00s =====================================================================
```

If I `rm testsinvalid/conftest.py`, I do see this.

## What I actually see

```
% PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 pytest --capture=no
ImportError while loading conftest '/[snip]/testsinvalid/conftest.py'.
testsinvalid/conftest.py:1: in <module>
    ???
E   Exception: Loading /[snip]/testsinvalid/conftest.py!
```

# Of note

If I `mv testsinvalid anything-without-the-prefix-tests`, the problem goes away.

**Comment by The-Compiler** [comment]
I can reproduce. No idea why it's happening, though.

**Comment by mattalexx** [comment]
@The-Compiler I narrowed it down to this block here:

https://github.com/pytest-dev/pytest/blob/1569fac603d9a50022e1474b494eebf970a2a3af/src/_pytest/config/__init__.py#L511-L513

But I don't know enough about the codebase to know what it should be doing.

**Comment by nicoddemus** [comment]
This has been added in d632a0d5c26c7c988a592e6b5a585685a576e36a, and while I agree it is surprising, it is a [documented feature](https://docs.pytest.org/en/stable/writing_plugins.html#plugin-discovery-order-at-tool-startup):

> if exists, load conftest.py and test*/conftest.py relative to the directory part of the first test path. 

So unless we want to somehow change this (including a deprecation period), we can close this as "worked as intended".

**Comment by The-Compiler** [comment]
That commit only moves the code around - it was initially added in 7bd60b5abb66e10cb107f79e412bfbb576e84b8b some 12 years ago:

> check and load test*/conftest.py early from anchors -
> this makes it a bit more convenient to have command line options
> available from a root directory of a project that does not
> directly contain a conftest.py

The `testpaths` option was only added 5 years later, in 854e603f84dde8345cd959ad302ddb5d281e67b2. It seems reasonable to me to use that setting (if set) rather than assuming `test*` blindly. Still probably a backwards-incompatible change though.

**Comment by bluetech** [comment]
Reading the code, using `testpaths` is not right here -- the testpaths themselves are iterated and used as anchors for `try_load_conftest` - i.e. the idea seems to be "`let's also consider test* subdirs`" *inside* of each testpath.

**Comment by bluetech** [comment]
Hmm, sorry, I might be confusing the variable name `testpaths` with the config `testpaths`. Disregard the above :)

**Comment by dhensen** [comment]
> This has been added in [d632a0d](https://github.com/pytest-dev/pytest/commit/d632a0d5c26c7c988a592e6b5a585685a576e36a), and while I agree it is surprising, it is a [documented feature](https://docs.pytest.org/en/stable/writing_plugins.html#plugin-discovery-order-at-tool-startup):
> 
> > if exists, load conftest.py and test*/conftest.py relative to the directory part of the first test path.
> 
> So unless we want to somehow change this (including a deprecation period), we can close this as "worked as intended".

Is this really expected behaviour? I'm in doubt so I'll share my setup using pytest 7.0.0:

```
├── lambda_layer
│   ├── layer.zip
│   └── python
│       ├── numpy
│       │   └── conftest.py     <--
├── pytest.ini
├── src
│   ├── mycodehere.py
├── tests
│   └── test_my_test_code_here.py

```

pytest.ini looks like this:
```
[tool:pytest]
addopts = -m 'not e2e'
testpaths = tests
```

When I execute `pytest` it fails, because numpy's conftest wants to import dependencies that I don't have installed. (I don't want to install those dependencies, I'm making a lambda layer, so those files are maybe there or maybe not, but it feels like it should not load all conftests under my repo root when I have specified testpaths)
Is this really expected behaviour?

> if exists, load conftest.py and test*/conftest.py relative to the directory part of the first test path

Isn't my first test path `tests` as specified via testpaths in my pytest.ini? I guess when [reading the manual regarding testpaths](https://docs.pytest.org/en/stable/reference/reference.html?highlight=testpaths#confval-testpaths) it does not really say anything about conftest loading behaviour. What to do with this?

**Comment by nicoddemus** [comment]
> Is this really expected behaviour?

@dhensen this seems like a different problem, because `lambda_layer` doesn't match `tests*`.

**Comment by nicoddemus** [comment]
@dhensen, in your `pytest.ini`, it should be only `[pytest]` in the header, `[tool:pytest]` is only used for `setup.cfg` files.

**Comment by JamieMcKernanKaizen** [comment]
I'm not sure why this is a feature though because I would expect running `pytest tests/` and using `testpaths = tests` when running `pytest` without any arguments to have the same behaviour, I thought that was the point of `testpaths`.

**Comment by nicoddemus** [comment]
> I'm not sure why this is a feature though because I would expect running pytest tests/ and using testpaths = tests when running pytest without any arguments to have the same behaviour, I thought that was the point of testpaths.

I agree, seems like that feature was introduced prior to `testpaths`.

I guess it is possible to deprecate the behavior: we can detect a `conftest.py` was matched via `test*/conftest.py`, and issue a deprecation warning.

**Comment by Zac-HD** [comment]
Closing as resolved: `testpaths` is now honored when determining which `test*/conftest.py` files to load at startup, implemented in #10988 (and refined by the regression fix in #11104 so unrelated sibling `test*` directories aren't scanned).
